src/storage/src/source/persist_source.rs

```diff
@@ -265,10 +303,54 @@ where
                 for (_idx, part) in buffer.drain(..) {
                     let (consumed_part, updates) = fetcher.fetch_leased_part(part.into()).await;
 
-                    let mut updates = updates
+                    let updates = updates
                         .expect("shard_id generated for sources must match across all workers");
 
-                    update_session.give_vec(&mut updates);
+                    // Apply as much logic to `updates` as we can, before we emit anything.
+                    let mut update_outputs = Vec::with_capacity(updates.len());
+                    for ((key, val), time, diff) in updates {
+                        if !until.less_equal(&time) {
+                            match (key, val) {
+                                (Ok(SourceData(Ok(row))), Ok(())) => {
+                                    if let Some(mfp) = &mut map_filter_project {
+                                        let arena = mz_repr::RowArena::new();
+                                        let mut datums_local = datum_vec.borrow_with(&row);
+                                        for result in mfp.evaluate(
+                                            &mut datums_local,
+                                            &arena,
+                                            time,
+                                            diff,
+                                            |time| !until.less_equal(time),
+                                            &mut row_builder,
+                                        ) {
+                                            match result {
+                                                Ok((row, time, diff)) => {
+                                                    if !until.less_equal(&time) {
```

Comment:
Are these here because temporal filters can futz with the timestamp?

Reply:
That's right.